### PR TITLE
Add required_roles to Lane and LaneFile

### PR DESCRIPTION
**Pull request by drasticactions**
This PR extends the Lane and LaneFile tables to include "required_roles". This is a comma separated text field that can include multiple roles inside a lane or a lane file. By having this, I could add ways inside of lanes and lane files a way to include individual roles (besides admins, who can see everything) the ability to clone and edit lanes they have access to, and only those. It's also done in such a way that we could extend this feature out further, should we wish to have it be more granular for certain features.

### CODE REVIEW

**Comment by duncanmak** [comment]
@rolfbjarne Can you take a look at this PR?

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs [hunk 1]
```diff
@@ -117,26 +129,27 @@ public void Logout (WebServiceLogin login)
 		}
 
 		[WebMethod]
-		public void CreateLanefile (WebServiceLogin login, int lane_id, string filename)
+		public void CreateLanefile (WebServiceLogin login, DBLane lane, string filename)
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: Don't change the signature of existing API (methods that have [WebMethod]), instead add a new method with the signature you want.

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs [hunk 2]
```diff
@@ -117,26 +129,27 @@ public void Logout (WebServiceLogin login)
...
 				throw new ArgumentException ("filename");
 
-			if (lane_id <= 0)
+			if (lane.id <= 0)
 				throw new ArgumentException ("lane_id");
 
 			using (DB db = new DB ()) {
-				VerifyUserInRole (db, login, Roles.Administrator);
-				db.Audit (login, "WebServices.CreateLaneFile (lane_id: {0}, filename: {1})", lane_id, filename);
+				VerifyUserInRoles (db, login, lane.required_roles, false);
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: This is not good, here you're taking `required_roles` as input, which means anybody can make a request and set `required_roles` to themselves.

You need to fetch the required roles from the database here, not accept the input.

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs [hunk 3]
```diff
@@ -1967,7 +1989,7 @@ public GetLaneFileForEditResponse GetLaneFileForEdit (WebServiceLogin login, int
 		public void EditLaneFile (WebServiceLogin login, DBLanefile lanefile)
 		{
 			using (DB db = new DB ()) {
-				VerifyUserInRole (db, login, Roles.Administrator);
+				VerifyUserInRoles (db, login, lanefile.required_roles, false);
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: This allows a user to add himself to the required roles.

File: scripts/database.sql [hunk 4]
```diff
@@ -179,7 +180,9 @@ CREATE TABLE Lanefile (
 	-- this is some simple change tracking
 	-- on every change a new Lanefile is stored, with the old contents and original_id referencing the real Lanefile
 	original_id    int        NULL REFERENCES Lanefile (id),
-	changed_date   timestamp  NULL -- the date the change was made	
+	changed_date   timestamp  NULL, -- the date the change was made
+
+	required_roles text NULL DEFAULT NULL	-- Roles required to access the Lanefile
```
Comment (rolfbjarne) [inline scripts/database.sql]: LaneFiles can be shared between lanes, so how does this work when a LaneFile is used by multiple lanes?

Also from what I can see there's no propagation of required roles from lanes to lane files, which means you'll have to edit the required roles manually for each file you want someone to be able to access. Is this the intended behavior?

Access management between shared resources get complicated quick 👿

**Comment by rolfbjarne** [comment]
I'm not sure I like the name `required roles`, wouldn't something like `additional roles` or `allowed roles` be better? After all you're not requiring a user to be a member of all the `required roles`, you're adding a list of users that can edit a lane _in addition to_ administrators.

File: scripts/database.sql (hunk 4 above)
Comment (drasticactions) [inline scripts/database.sql]: When a user creates a new file in a lane, it copies over the `required _roles` from the lane it was created from (so the user can still see it). For a LaneFile that has lesser permissions than what the current user has in the lane they are in, they would be unauthorized. Someone with role access or an admin does need to go in and add the user role to it. This is indented behavior, because if you want to have a permission system for who can see and access a file, you should be explicit about it and add them in. This goes for the file history too. When you change the required roles for the file, users can go to the file history and see there were changes, but can't access them (since they didn't have access to them at that time). Also for admins, nothing changes. They get full access to every file or lane regardless of the roles in those fields.

For LaneFiles, they only need to do this once, since the `required_roles` for the LaneFile are separate from the lane. So I would think for shared resources this would be fine. 

I

**Comment by drasticactions** [comment]
Those all basically sound the exact same to me, but I'm cool with `allowed roles`.

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs (hunk 2 above)
Comment (drasticactions) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: When I wrote it, I thought that these calls were only being made from inside EditLane, which requires you to have the permissions to get on the page at all. If this is an API call that can be access from outside of that scope, then yes, it won't work.

`[WebMethod]` does not really say "API Access" as much as `WebAPI`. I blame WebForms ;)

Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: Anything with a `[WebMethod]` is accessible from outside (it's web service).

**Comment by drasticactions** [comment]
@rolfbjarne I added the original endpoints back to WebServices (which require Admin access to use), and added new one which take in the `lane_id`, reload the lane in the method and check for roles with that new instance. `required_roles` is now `additional_roles` as well.

For `EditLane` and `EditLaneFile`, you need to have permissions on the old file or lane in order to edit the new file or lane.

Could you check through this again to see if I missed anything?

File: MonkeyWrench.Web.UI/EditHost.aspx.cs [hunk 5]
```diff
@@ -78,22 +82,26 @@ protected override void OnInit (EventArgs e)
 			}
 
 			if (!string.IsNullOrEmpty (disable) && int.TryParse (disable, out id)) {
-				Utils.LocalWebService.SwitchHostEnabledForLane (Master.WebServiceLogin, id, response.Host.id);
+				DBLane lane = Utils.LocalWebService.FindLane (Master.WebServiceLogin, id, null).lane;
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.UI/EditHost.aspx.cs]: Here you fetch the lane using the id, just to only use lane.id the next line. That looks redundant.

File: MonkeyWrench.Web.UI/EditLane.aspx.cs [hunk 6]
```diff
@@ -129,22 +135,22 @@ protected override void OnInit (EventArgs e)
 				break;
 			case "editCommandFilename":
 				if (int.TryParse (command_id, out id))
-					Utils.LocalWebService.EditCommandFilename (Master.WebServiceLogin, id, Request ["filename"]);
+					Utils.LocalWebService.EditCommandFilename (Master.WebServiceLogin, lane.id, id, Request ["filename"]);
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.UI/EditLane.aspx.cs]: These changes (to use lane.id instead of id) are not needed anymore now I think

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs [hunk 7]
```diff
@@ -126,17 +138,19 @@ public void CreateLanefile (WebServiceLogin login, int lane_id, string filename)
 				throw new ArgumentException ("lane_id");
 
 			using (DB db = new DB ()) {
-				VerifyUserInRole (db, login, Roles.Administrator);
-				db.Audit (login, "WebServices.CreateLaneFile (lane_id: {0}, filename: {1})", lane_id, filename);
+				var lane = DBLane_Extensions.Create (db, lane_id);
+				VerifyUserInRoles (db, login, lane.additional_roles, false);
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: This allows anybody to create the lanefile if `lane.additional_roles` is empty.

I think you need to create a list where you make sure Administrator is, and then add any additional roles to that list.

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs [hunk 8]
```diff
@@ -126,17 +138,19 @@ public void CreateLanefile (WebServiceLogin login, int lane_id, string filename)
...
 				DBLanefile file = new DBLanefile ();
 				file.name = filename;
 				file.contents = "#!/bin/bash -ex\n\n#Your commands here\n";
 				file.mime = "text/plain";
+				file.additional_roles = lane.additional_roles;
 				file.Save (db);
 
 				DBLanefiles lanefile = new DBLanefiles ();
-				lanefile.lane_id = lane_id;
+				lanefile.lane_id = lane.id;
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: Redundant change.

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs [hunk 9]
```diff
@@ -148,10 +162,11 @@ public void CreateLanefile (WebServiceLogin login, int lane_id, string filename)
...
 		{
 			using (DB db = new DB ()) {
-				VerifyUserInRole (db, login, Roles.Administrator);
-				db.Audit (login, "WebServices.AttachFileToLane (lane_id: {0}, lanefile_id: {1})", lane_id, lanefile_id);
+				var lane = DBLane_Extensions.Create (db, lane_id);
+				VerifyUserInRoles (db, login, lane.additional_roles, false);
+				db.Audit (login, "WebServices.AttachFileToLane (lane_id: {0}, lanefile_id: {1})", lane.id, lanefile_id);
 				DBLanefiles lanefile = new DBLanefiles ();
-				lanefile.lane_id = lane_id;
+				lanefile.lane_id = lane.id;
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: Redundant change.

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs [hunk 10]
```diff
@@ -148,10 +162,11 @@ public void CreateLanefile (WebServiceLogin login, int lane_id, string filename)
 		public void AttachFileToLane (WebServiceLogin login, int lane_id, int lanefile_id)
 		{
 			using (DB db = new DB ()) {
-				VerifyUserInRole (db, login, Roles.Administrator);
-				db.Audit (login, "WebServices.AttachFileToLane (lane_id: {0}, lanefile_id: {1})", lane_id, lanefile_id);
+				var lane = DBLane_Extensions.Create (db, lane_id);
+				VerifyUserInRoles (db, login, lane.additional_roles, false);
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: Same here: make sure Administrator is in the list of roles.

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs [hunk 11]
```diff
@@ -161,18 +176,30 @@ public void AttachFileToLane (WebServiceLogin login, int lane_id, int lanefile_i
 		public void DeattachFileFromLane (WebServiceLogin login, int lane_id, int lanefile_id)
 		{
 			using (DB db = new DB ()) {
-				VerifyUserInRole (db, login, Roles.Administrator);
-				db.Audit (login, "WebServices.DeattachFileFromLane (lane_id: {0}, lanefile_id: {1})", lane_id, lanefile_id);
+				var lane = DBLane_Extensions.Create (db, lane_id);
+				VerifyUserInRoles (db, login, lane.additional_roles, false);
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: Make sure Administrator is in the list of roles.

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs [hunk 12]
```diff
@@ -161,18 +176,30 @@ public void AttachFileToLane (WebServiceLogin login, int lane_id, int lanefile_i
...
+				db.Audit (login, "WebServices.DeattachFileFromLane (lane_id: {0}, lanefile_id: {1})", lane.id, lanefile_id);
 				using (IDbCommand cmd = db.CreateCommand ()) {
 					cmd.CommandText = "DELETE FROM Lanefiles WHERE lane_id = @lane_id AND lanefile_id = @lanefile_id;";
-					DB.CreateParameter (cmd, "lane_id", lane_id);
+					DB.CreateParameter (cmd, "lane_id", lane.id);
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: Redundant change.

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs [hunk 13]
```diff
@@ -161,18 +176,30 @@ public void AttachFileToLane (WebServiceLogin login, int lane_id, int lanefile_i
...
 				}
 			}
 		}
 
 		[WebMethod]
+		public void EditCommand (WebServiceLogin login, DBCommand command, int lane_id)
+		{
+			using (DB db = new DB ()) {
+				var lane = DBLane_Extensions.Create (db, lane_id);
+				VerifyUserInRoles (db, login, lane.additional_roles, false);
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: Make sure Administrator is in the list of roles.

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs [hunk 14]
```diff
@@ -183,6 +210,19 @@ public void EditCommand (WebServiceLogin login, DBCommand command)
 		}
 
 		[WebMethod]
+		public void EditCommandFilename (WebServiceLogin login, int lane_id, int command_id, string filename)
+		{
+			using (DB db = new DB ()) {
+				var lane = DBLane_Extensions.Create (db, lane_id);
+				VerifyUserInRoles (db, login, lane.additional_roles, false);
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: Make sure Administrator is in the list of roles.

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs [hunk 15]
```diff
@@ -195,6 +235,19 @@ public void EditCommandFilename (WebServiceLogin login, int command_id, string f
 		}
 
 		[WebMethod]
+		public void EditCommandSequence (WebServiceLogin login, int lane_id, int command_id, int sequence)
+		{
+			using (DB db = new DB ()) {
+				var lane = DBLane_Extensions.Create (db, lane_id);
+				VerifyUserInRoles (db, login, lane.additional_roles, false);
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: Make sure Administrator is in the list of roles.

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs (hunk 11 above)
Comment (drasticactions) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: VerifyUserInRoles automatically adds Administrator to the `additional_roles` to the list.

Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: I looked for that, but I couldn't find it?

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs [hunk 16]
```diff
@@ -161,18 +176,30 @@ public void AttachFileToLane (WebServiceLogin login, int lane_id, int lanefile_i
...
-					DB.CreateParameter (cmd, "lane_id", lane_id);
+					DB.CreateParameter (cmd, "lane_id", lane.id);
 					DB.CreateParameter (cmd, "lanefile_id", lanefile_id);
 					cmd.ExecuteNonQuery ();
 				}
 			}
 		}
 
 		[WebMethod]
+		public void EditCommand (WebServiceLogin login, DBCommand command, int lane_id)
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: Web services doesn't support overloading methods, so please use a different name (and the same goes for all the other new methods).

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs [hunk 17]
```diff
@@ -279,6 +410,18 @@ public void SwitchCommandNonFatal (WebServiceLogin login, int command_id)
 		}
 
 		[WebMethod]
+		public void SwitchCommandAlwaysExecute (WebServiceLogin login, int lane_id, int command_id)
+		{
+			using (DB db = new DB ()) {
+				var lane = DBLane_Extensions.Create (db, lane_id);
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: You're not verifying that `command_id`'s lane is `lane_id`, which means someone can call this method for all command_ids as long as they have permission to change one lane.

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs [hunk 18]
```diff
@@ -2205,10 +2438,10 @@ public int GetUploadPort ()
 		}
 
 		[WebMethod]
-		public int AddEnvironmentVariable (WebServiceLogin login, int? lane_id, int? host_id, string name, string value)
+		public int AddEnvironmentVariable (WebServiceLogin login, int? lane_id, int? host_id, string name, string value, string[] requiredRoles)
 		{
 			using (DB db = new DB ()) {
-				VerifyUserInRole (db, login, Roles.Administrator);
+				VerifyUserInRoles (db, login, requiredRoles, false);
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: Here you're still using the input to validate against.

**Comment by drasticactions** [comment]
@rolfbjarne Made more review changes. I've changed the names of the WebMethods I've added to WebServices. 

When we started doing this I was not intending on adding more API calls, but I left in the WebMethod attribute because it _could_ be useful to have and I don't see a reason to keep them internal. But I'm fine removing those attributes and refactoring it more if need be.

File: MonkeyWrench.Web.UI/EditHost.aspx.cs [hunk 19]
```diff
@@ -82,12 +82,12 @@ protected override void OnInit (EventArgs e)
 				redirect = true;
 			}
 
-			if (!string.IsNullOrEmpty (enable) && int.TryParse (enable, out id)) {
+			if (!string.IsNullOrEmpty (enable) && int.TryParse (enable, out id)) {
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.UI/EditHost.aspx.cs]: This looks like unnecessary whitespace change (and the line below as well).

File: MonkeyWrench.Web.UI/EditLane.aspx.cs [hunk 20]
```diff
@@ -29,6 +29,8 @@ public partial class EditLane : System.Web.UI.Page
 {
 	private DBLane lane;
 
+	private string[] requiredRoles;
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.UI/EditLane.aspx.cs]: Could you rename this one too to `additionalRoles`?

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs [hunk 21]
```diff
@@ -55,6 +55,21 @@ private void VerifyUserInRole (DB db, WebServiceLogin login, string role, bool @
...
+			string[] r = (roles == null) ? new string[0] : roles.Split(',');
+
+			Authentication.VerifyUserInRoles(Context, db, login, r, @readonly);
+		}
+
+		private void VerifyUserInRoles (DB db, WebServiceLogin login, string[] roles, bool @readonly)
+		{
+			// Administrator should be given access to everything.
+			// So any call to VerifyUserInRoles should add Administrator to that list.
+			roles.Concat(new string[] { Roles.Administrator });
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: Arrays are immutable, so this will not modify `roles`, instead it will return a new array.

Instead do this:

    roles = roles.Concat (...);

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs [hunk 22]
```diff
@@ -55,6 +55,21 @@ private void VerifyUserInRole (DB db, WebServiceLogin login, string role, bool @
 			Authentication.VerifyUserInRole (Context, db, login, role, @readonly);
 		}
 
+		private void VerifyUserInRoles(DB db, WebServiceLogin login, string roles, bool @readonly)
+		{
+			string[] r = (roles == null) ? new string[0] : roles.Split(',');
+
+			Authentication.VerifyUserInRoles(Context, db, login, r, @readonly);
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: You're not adding `Administrator` to the list here (and you're not calling the overload just below either).

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs [hunk 23]
```diff
@@ -394,17 +603,30 @@ public void AddDependencyToLane (WebServiceLogin login, int lane_id, int depende
...
 		}
 
 		[WebMethod]
+		public void EditLaneDependencyFilenameInLane (WebServiceLogin login, int lane_id, int lanedependency_id, string filename)
+		{
+			using (DB db = new DB ()) {
+				var lane = DBLane_Extensions.Create (db, lane_id);
+				VerifyUserInRoles (db, login, lane.additional_roles, false);
+				DBLaneDependency dep = DBLaneDependency_Extensions.Create (db, lanedependency_id);
+				dep.filename = filename;
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: Here too you need to verify that the lanedependency is from lane_id.

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs [hunk 24]
```diff
@@ -649,6 +894,16 @@ public GetLaneForEditResponse GetLaneForEdit (WebServiceLogin login, int lane_id
 					response.Lane = response.Lanes.Find ((l) => l.lane == lane);
 				}
 
+				// Get the required roles for the lane. Administrator can access all lanes.
+				var requiredRoles = new List<string> ();
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: Please rename this to `additionalRoles`.

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs [hunk 25]
```diff
@@ -2221,6 +2495,16 @@ public int AddEnvironmentVariable (WebServiceLogin login, int? lane_id, int? hos
 		}
 
 		[WebMethod]
+		public void EditEnvironmentVariableInLane (WebServiceLogin login, DBEnvironmentVariable variable, int lane_id)
+		{
+			using (DB db = new DB ()) {
+				var lane = DBLane_Extensions.Create (db, lane_id);
+				VerifyUserInRoles (db, login, lane.additional_roles, false);
+				variable.Save (db);
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: You need to verify that the environment variable's lane is lane_id.

File: MonkeyWrench.Web.WebService/WebServices.asmx.cs [hunk 26]
```diff
@@ -3098,6 +3393,26 @@ public WebServiceResponse AddLaneNotification (WebServiceLogin login, int lane_i
...
+		public WebServiceResponse RemoveLaneNotificationForLane (WebServiceLogin login, int id, int lane_id)
+		{
+			WebServiceResponse response = new WebServiceResponse ();
+
+			using (DB db = new DB ()) {
+				var lane = DBLane_Extensions.Create (db, lane_id);
+				VerifyUserInRoles (db, login, lane.additional_roles, false);
+
+				using (IDbCommand cmd = db.CreateCommand ()) {
+					cmd.CommandText = "DELETE FROM LaneNotification WHERE id = @id;";
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/WebServices.asmx.cs]: You need to verify that the notification is from lane_id (should be just a matter of adding the condition to the sql statement).

File: MonkeyWrench.Web.WebService/Authentication.cs [hunk 27]
```diff
@@ -127,13 +127,24 @@ private static void VerifyAnonymousAllowed()
...
+
+		public static void VerifyUserInRoles (HttpContext Context, DB db, WebServiceLogin login, string[] roles, bool @readonly)
+		{
 			WebServiceResponse dummy = new WebServiceResponse ();
 			Authenticate (Context, db, login, dummy, @readonly);
 
-			if (!dummy.IsInRole (role)) {
-				log.InfoFormat ("The user '{0}' has the roles '{1}', and requested role is: {2}", login.User, dummy.UserRoles == null ? "<null>" : string.Join (",", dummy.UserRoles), role);
-				throw new UnauthorizedException ("You don't have the required permissions.");
+			foreach (var role in roles) {
```
Comment (rolfbjarne) [inline MonkeyWrench.Web.WebService/Authentication.cs]: What happens here if one of the roles is an empty string?

I think that can happen if someone just puts in a comma for the field when editing the lane.

**Comment by rolfbjarne** [comment]
I don't think it's necessary to expose the new methods for now, so maybe just remove the [WebMethod] attribute from them  (no reason to make the attack surface bigger unless we actually need the functionality).

File: MonkeyWrench.Web.WebService/Authentication.cs (hunk 27 above)
Comment (drasticactions) [inline MonkeyWrench.Web.WebService/Authentication.cs]: Good point. If it's blank it should be okay, but if someone mistypes with an extra comma there could be problems. There should at least be a empty/null check for that case. I'll add it.

**Comment by drasticactions** [comment]
Okay, I should have removed all of the WebMethod attributes for the new ones I added, and kept the ones that were there to start with.
